### PR TITLE
Save FFI module in fem form.

### PR DESCRIPTION
**Pull request by bpachev**
This is a relatively minor change to make the ffi module returned by ffcx available in the Form class, in addition to the ufcx_form Python object and the code. The reason for this PR is to support my add-on package providing GPU accelerated assembly on top of dolfinx (https://github.com/bpachev/cuda-dolfinx). I need a way to pass the ufcx_form pointer to C, similar to what is done in the instantiation of the cpp Form class, e.g:

```
        f = ftype(
            module.ffi.cast("uintptr_t", module.ffi.addressof(ufcx_form)),
            V,
            coeffs,
            constants,
            subdomains,
            _entity_maps,
            mesh,
        )
```

The module variable is discarded after fe.form returns. I'm aware that it is possible to find the module by replicating the cache search logic within ffcx, but feel like saving module in Form is much simpler and stabler. I apologize in advance if this change isn't needed. If there's a simpler solution, I'm all ears :).

### CODE REVIEW

File: python/dolfinx/fem/forms.py [hunk 1]
```diff
@@ -47,6 +47,7 @@ def __init__(
         ],
         ufcx_form=None,
         code: typing.Optional[str] = None,
+        module=None,
```
Comment (garth-wells) [inline python/dolfinx/fem/forms.py:50]: Add a type hint for `module`.

File: python/dolfinx/fem/forms.py [hunk 2]
```diff
@@ -75,6 +78,11 @@ def code(self) -> typing.Union[str, None]:
         """C code strings"""
         return self._code
 
+    @property
+    def module(self):
```
Comment (garth-wells) [inline python/dolfinx/fem/forms.py:82]: Add return type hint.

**Review by garth-wells — DISMISSED** [review]
Looks good - just needs some type hints.

File: python/dolfinx/fem/forms.py (hunk 1 above)
Comment (bpachev) [reply python/dolfinx/fem/forms.py]: Done

File: python/dolfinx/fem/forms.py (hunk 2 above)
Comment (bpachev) [reply python/dolfinx/fem/forms.py]: Done

**Comment by jhale** [comment]
Can you ruff check and ruff format this? Then we will merge.

File: python/dolfinx/fem/forms.py [hunk 3]
```diff
@@ -47,6 +48,7 @@ def __init__(
         ],
         ufcx_form=None,
         code: typing.Optional[str] = None,
+        module: typing.Optional[types.ModuleType] = None,
```
Comment (jhale) [inline python/dolfinx/fem/forms.py:51]: Is types.ModuleType correct? Not what I expected.

Comment (bpachev) [reply python/dolfinx/fem/forms.py:51]: The module object is created in ffcx using importlib.util.module_from_spec. According to
https://docs.python.org/3/library/types.html#types.ModuleType, types.ModuleType appears to be the correct type for Python modules.

**Comment by bpachev** [comment]
Fixed the linting issue and addressed the comment about the type.

**Pull request merged**